src/Prefetch.cpp

```diff
@@ -27,13 +27,48 @@ const Definition &get_stage_definition(const Function &f, int stage_num) {
     return f.update(stage_num - 1);
 }
 
+// Collect the bounds of all the externally referenced buffers in a stmt.
+class CollectExternalBufferBounds : public IRVisitor {
+public:
+    map<string, Box> buffers;
+
+    using IRVisitor::visit;
+
+    void add_buffer_bounds(const string &name, Buffer<> image, Parameter param, int dims) {
+        Box b;
+        for (int i = 0; i < dims; ++i) {
+            string dim_name = std::to_string(i);
+            Expr buf_min_i = Variable::make(Int(32), name + ".min." + dim_name,
+                                            image, param, ReductionDomain());
+            Expr buf_extent_i = Variable::make(Int(32), name + ".extent." + dim_name,
+                                               image, param, ReductionDomain());
+            Expr buf_max_i = buf_min_i + buf_extent_i - 1;
+            b.push_back(Interval(buf_min_i, buf_max_i));
+        }
+        buffers.emplace(name, b);
+    }
+
+    void visit(const Call *op) {
+        IRVisitor::visit(op);
+        add_buffer_bounds(op->name, op->image, op->param, (int)op->args.size());
+    }
+
+    void visit(const Variable *op) {
+        if (op->param.defined() &&
+            op->param.is_buffer()) {
```

Comment:
Minor nit: this easily fits on one line and multi-line conditions are a bit harder to read quickly.